cpp/include/Ice/PluginFactory.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) ZeroC, Inc.
+
+#ifndef ICE_PLUGIN_FACTORY_H
+#define ICE_PLUGIN_FACTORY_H
+
+#include "CommunicatorF.h"
+#include "Config.h"
+#include "Ice/BuiltinSequences.h"
+#include "Plugin.h"
+
+#include <string>
+
+namespace Ice
+{
+    class Plugin;
+
+    /// A plug-in factory function is a C function responsible for creating an Ice plug-in.
+    /// @param communicator The communicator in which the plug-in will be installed.
+    /// @param name The name assigned to the plug-in.
+    /// @param args Additional arguments included in the plug-in's configuration.
+    /// @return The new plug-in object.
+    using PluginFactoryFunc = Plugin* (*)(const CommunicatorPtr& communicator,
+                                          const std::string& name,
+                                          const StringSeq& args);
+
+    /// Represents a plug-in factory.
+    /// @see InitializeData::pluginFactories
+    struct PluginFactory
+    {
+        /// The default and preferred name for plug-ins created by this factory.
+        std::string pluginName;
+
+        /// The factory function.
+        PluginFactoryFunc factoryFunc;
+    };
+
+#ifndef ICE_PLUGIN_FACTORY_DECLSPEC_IMPORT
+#    define ICE_PLUGIN_FACTORY_DECLSPEC_IMPORT ICE_DECLSPEC_IMPORT
+#endif
+
+    //
+    // Checking for the API_EXPORTS macro is necessary to prevent inconsistent DLL linkage errors on Windows.
+    //
+
+#if !defined(ICE_API_EXPORTS) || defined(ICE_DOXYGEN)
```

Comment:
Can you explain the logic for the export/import macros usage here?

Reply:
It's only relevant on Windows, so maybe we should rework these since we don't support "_min" builds on Windows, and as a result, udp/ws are always built-in.

Let's take IceDiscovery on Windows. We want applications to be able to "import" IceDiscovery (with ICE_DECLSPEC_IMPORT), however IceDiscovery itself should not import this function (it needs to export it).

Also, `ICE_DISCOVERY_API` is not defined when including Ice/Ice.h. That's why we rely on this roundabout `!defined(ICE_DISCOVERY_API_EXPORTS)`.

Reply:
For the built-in Ice plug-ins, it would be cleaner to just use ICE_API.